tests/kernel/msgq/msgq_usage/src/main.c

```diff
@@ -0,0 +1,254 @@
+			client = (struct k_msgq *)service_data[1];
+			service_data[0] = SERVICE2_RUNNING;
+			k_msgq_put(client, service_data, K_NO_WAIT);
+		}
+	}
+
+	/* inform services manager */
+	service_data[0] = SERVICE_QUIT;
+	service_data[1] = (unsigned long)&service2_msgq;
+	k_msgq_put(&manager_q, service_data, K_NO_WAIT);
+}
+
+static void register_service(void)
+{
+
+	int pri = k_thread_priority_get(k_current_get());
+
+	service1_run = true;
+	k_thread_create(&service1, service1_stack, STACK_SIZE, service1_entry,
+			NULL, NULL, NULL, pri, 0, K_NO_WAIT);
+
+	service2_run = true;
+	k_thread_create(&service2, service2_stack, STACK_SIZE, service2_entry,
+			NULL, NULL, NULL, pri, 0, K_NO_WAIT);
+}
+
+static void client_entry(void *p1, void *p2, void *p3)
+{
+	unsigned long client_data[2];
+	unsigned long service_data[2];
+	struct k_msgq *service1q;
+	struct k_msgq *service2q;
+	bool query_service = false;
+	int ret;
+
+
+	k_msgq_init(&client_msgq, (char *)client_buf,
+		    sizeof(unsigned long) * 2, 2);
+	client_data[0] = QUERRY_SERVICE;
+	client_data[1] = (unsigned long)&client_msgq;
+
+	/* query services */
+	k_msgq_put(&manager_q, client_data, K_NO_WAIT);
+	ret = k_msgq_get(&client_msgq, service_data, K_FOREVER);
+	zassert_equal(ret, 0, NULL);
+
+	service1q = (struct k_msgq *)service_data[0];
+	service2q = (struct k_msgq *)service_data[1];
+	/* all services should be running */
+	zassert_equal(service1q, &service1_msgq, NULL);
+	zassert_equal(service2q, &service2_msgq, NULL);
+
+	while (1) {
+		/* service might quit */
+		if (query_service) {
+			client_data[0] = QUERRY_SERVICE;
+			client_data[1] = (unsigned long)&client_msgq;
+			k_msgq_put(&manager_q, client_data, K_NO_WAIT);
+			k_msgq_get(&client_msgq, service_data, K_FOREVER);
+			service1q = (struct k_msgq *)service_data[0];
+			service2q = (struct k_msgq *)service_data[1];
+			query_service = false;
+		}
+
+		if (!service1q && !service2q) {
+			break;
+		}
+
+		client_data[0] = GET_SERVICE;
+		client_data[1] = (unsigned long)&client_msgq;
+
+		if (service1q) {
+			k_msgq_put(service1q, client_data, K_NO_WAIT);
+			ret = k_msgq_get(&client_msgq, service_data, TIMEOUT);
+			if (!ret) {
+				zassert_equal(service_data[0],
+					      SERVICE1_RUNNING, NULL);
+			} else {
+				/* service might down, query in next loop */
+				query_service = true;
+			}
+		}
+
+		if (service2q) {
+			k_msgq_put(service2q, client_data, K_NO_WAIT);
+			ret = k_msgq_get(&client_msgq, service_data, TIMEOUT);
+			if (!ret) {
+				zassert_equal(service_data[0],
+					      SERVICE2_RUNNING, NULL);
+			} else {
+				query_service = true;
+			}
+		}
+	}
+}
+
+static void start_client(void)
+{
+
+	int pri = k_thread_priority_get(k_current_get());
+
+	k_thread_create(&client, client_stack, STACK_SIZE, client_entry,
+			NULL, NULL, NULL, pri, 0, K_NO_WAIT);
+}
+
+void test_msgq_usage(void)
+{
+	start_service_manager();
+	register_service();
+	start_client();
+	k_msleep(500);
+	service1_run = false;
```

Comment:
It is accessed/set in multi-threads. I think we need to a lock for it.

Reply:
updated

Reply:
@andyross test cases in tests/kernel/msgq/msgq_api are module test cases, this patch provides an integration test case, yes it looks like a sample, but I think I can write a sample later.

Reply:
@mengxianglinx , the comment above has nothing to do with the thread. I guess that you start another thread for it.